Docs for Collaborative Training (#12996)
PyTorchLightning/pytorch-lightning: docs/source/strategies/collaborative_training_expert.rst

:orphan:

.. _collaborative_training_expert:

Training on unreliable mixed GPUs across the internet (Expert)
==============================================================

Using Compression to Optimize communications
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Below are some ways to reduce communication when training collaboratively. As the model sizes get larger, bottlenecks in communication become more apparent.

Reviewer: ```suggestion
Below are some ways to reduce communication when training collaboratively. As the size of your model increase, bottlenecks in communication become more apparent.
```

Compress Gradients & State
""""""""""""""""""""""""""

> Hivemind allows you to compress gradients and states before sending them to other machines. This helps reduce the communication overhead substantially when training across the internet.

Author: I just added a note to mention this may affect convergence if going from Float32 -> Float16. 

I haven't done enough research around intra-node performance, but hivemind always uses the network adapter (local DDP support is a later thing to work on).

> Below we enable Float16 compression, which compresses gradients and states to Float16 before sending it to other machines.

Reviewer: ```suggestion
Below, we enable Float16 compression, which compresses gradients and states to Float16 before sending it to other machines.
```

.. code-block:: python

    from hivemind import Float16Compression
    import pytorch_lightning as pl
    from pytorch_lightning.strategies import CollaborativeStrategy

    trainer = pl.Trainer(
        strategy=CollaborativeStrategy(
            target_batch_size=target_batch_size,
            grad_compression=Float16Compression(),
            state_averaging_compression=Float16Compression(),
        ),
        accelerator="gpu",
        devices=1,
    )

> A slightly more advanced scheme is dynamic compression based on the size of values. Below we enable 8-bit quantization for large numbers, and Float16 compression for small values.

Reviewer: ```suggestion
A slightly more advanced scheme is dynamic compression based on value size. Below, we enable 8-bit quantization for large numbers, and Float16 compression for small values.
```

> Size Adaptive Compression has been used in a variety of Hivemind applications and has shown success.

Author: extended the note!

    compression = SizeAdaptiveCompression(
        threshold=2 ** 16 + 1, less=Float16Compression(), greater_equal=Uniform8BitQuantization()
    )
    trainer = pl.Trainer(
        strategy=CollaborativeStrategy(
            target_batch_size=target_batch_size,
            grad_compression=compression,
            state_averaging_compression=compression,
        ),
        accelerator="gpu",
        devices=1,
    )

PowerSGD
""""""""

`PowerSGD <https://arxiv.org/abs/1905.13727>`_ is a technique to reduce distributed communication of gradients across processes.
In short, PowerSGD uses a low-rank approximation to compress gradients before all reducing.

.. note::
    PowerSGD may impact convergence, however, it is worth trying as it can substantially reduce the communication between processes.

.. code-block:: python

    import pytorch_lightning as pl
    from pytorch_lightning.strategies import CollaborativeStrategy
    from functools import partial
    from hivemind.optim.power_sgd_averager import PowerSGDGradientAverager

    trainer = pl.Trainer(
        strategy=CollaborativeStrategy(
            target_batch_size=8192,

>             grad_averager_factory=partial(PowerSGDGradientAverager, averager_rank=32, min_compression_ratio=0.5),

Author: That's a great question, I was given these values but I wonder if there are ways to pick them other than trying different values.

cc @justheuristic

>             grad_averager_factory=partial(PowerSGDGradientAverager, averager_rank=32, min_compression_ratio=0.5),

Author: Will address in a future PR, we might be able to automate some of these variables.

:orphan:

.. _collaborative_training_expert:

Training on unreliable mixed GPUs across the internet (Expert)
==============================================================

> Using Compression to Optimize communications

Reviewer 2: nit 🐛 
```suggestion
Using Compression to Optimize Communications
```
